### PR TITLE
User entity: UserServlet

### PR DESCRIPTION
**Pull request by bori00**
The UserServlet is a webservlet which handles requests to store or to retrieve the currently logged in user's data. 

- doGet() returns as a response a User entity, containing all the information saved about the current user in the database
- doPost() saves the data sent in the request in the database, as a User entity

UserServletTest contains some tests which verify if the UserServlet behaves as expected. These tests rely on mock requests and responses and some fake data for the UserService.

### CODE REVIEW

**Comment by bori00** [comment]
This PR relies in the code sent to the "User entity: class" and "User entity: database" PR, which were not merged to the master yet.

File: src/main/java/com/google/sps/servlets/UserServlet.java [hunk 1]
```diff
@@ -14,14 +14,74 @@
...
+        userRepository = UserRepositoryFactory.getUserRepository(RepositoryType.DATASTORE);
+        userService = UserServiceFactory.getUserService();
+    }
+
+     /** Saves the recently submitted userdata(updates it if the user already has some data saved). */
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException { 
+        User user = getUserFromRequest(request);
+        userRepository.saveUser(user);
+        response.sendRedirect("/portfolio.html");
```
Comment (denniswillie) [inline src/main/java/com/google/sps/servlets/UserServlet.java]: can you change this to `index.html`? thanks :)

File: src/main/java/com/google/sps/servlets/UserServlet.java [hunk 2]
```diff
@@ -14,14 +14,74 @@
...
+        response.setContentType("application/json;");
+        response.getWriter().println(convertToJsonUsingGson(user));
+    }
+
+    private User getUserFromRequest(HttpServletRequest request) {
+        String id = userService.getCurrentUser().getUserId();
+        String email = userService.getCurrentUser().getEmail();
+        User.Builder newUserBuilder = new User.Builder(id, email);
+        String name = request.getParameter(NAME_INPUT);
+        if (!name.equals("")) {
```
Comment (denniswillie) [inline src/main/java/com/google/sps/servlets/UserServlet.java]: just out of curiosity, if the entered name is a double space, will `!name.equals("")` accommodate that?

File: src/test/java/com/google/sps/UserServletTest.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    @After
+    public void tearDown() {
+        helper.tearDown();
+    }
+ 
+    @Test
+    public void doPostDoGet_existingUserWithAllDataSet_returnsUserWithAllData() throws IOException, ServletException {
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        HttpServletResponse response = mock(HttpServletResponse.class);
```
Comment (denniswillie) [inline src/test/java/com/google/sps/UserServletTest.java]: mocking should be done in the setup method.

File: src/test/java/com/google/sps/UserServletTest.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        userServlet = new UserServlet();
+    }
+
+    @After
+    public void tearDown() {
+        helper.tearDown();
+    }
+ 
+    @Test
+    public void doPostDoGet_existingUserWithAllDataSet_returnsUserWithAllData() throws IOException, ServletException {
```
Comment (denniswillie) [inline src/test/java/com/google/sps/UserServletTest.java]: Split this method to 2 different methods, one for `doPost` and one for `doGet`.

File: src/test/java/com/google/sps/UserServletTest.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        User resultUser = gson.fromJson(sw.toString(), User.class);
+        assertEquals(ID, resultUser.getId());
+        assertEquals(EMAIL, resultUser.getEmail());
+        assertEquals(NAME, resultUser.getName());
+        assertTrue(resultUser.portfolioIsPublic());
+        assertEquals(SELF_INTRODUCTION, resultUser.getSelfIntroduction());
+    }
+
+    @Test
+    public void doPostDoGet_existingUserWithNoDataSet_returnsUserWithMissingData() throws IOException, ServletException {
```
Comment (denniswillie) [inline src/test/java/com/google/sps/UserServletTest.java]: Split this method to 2 different methods, one for doPost and one for doGet.

File: src/main/java/com/google/sps/servlets/UserServlet.java (hunk 2 above)
Comment (bori00) [reply src/main/java/com/google/sps/servlets/UserServlet.java]: This method doesn't check the validity of the name entered by the user, I think, that should be done on client side. Here I check only if the user entered something. The user shouldn't be able to submit the form if some inputs are invalid.

File: src/main/java/com/google/sps/servlets/UserServlet.java (hunk 1 above)
Comment (bori00) [reply src/main/java/com/google/sps/servlets/UserServlet.java]: Changed

File: src/test/java/com/google/sps/UserServletTest.java (hunk 3 above)
Comment (bori00) [reply src/test/java/com/google/sps/UserServletTest.java]: Done.

File: src/test/java/com/google/sps/UserServletTest.java (hunk 4 above)
Comment (bori00) [reply src/test/java/com/google/sps/UserServletTest.java]: Done.

File: src/test/java/com/google/sps/UserServletTest.java (hunk 5 above)
Comment (bori00) [reply src/test/java/com/google/sps/UserServletTest.java]: Done.

File: src/main/java/com/google/sps/servlets/UserServlet.java (hunk 2 above)
Comment (denniswillie) [reply src/main/java/com/google/sps/servlets/UserServlet.java]: okay

**Comment by denniswillie** [comment]
LGTM!

File: src/main/java/com/google/sps/servlets/UserServlet.java [hunk 6]
```diff
@@ -14,14 +14,74 @@
...
```
Comment (tlheng) [inline src/main/java/com/google/sps/servlets/UserServlet.java]: !name.isEmpty()

File: src/test/java/com/google/sps/UserServletTest.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+@RunWith(JUnit4.class)
+public final class UserServletTest {
+    private static final String ID = "userid";
+    private static final String EMAIL = "user@gmail.com";
+    private static final String NAME = "username";
+    private static final String SELF_INTRODUCTION = "I am the user";
+    private static final String IMG_URL = "/img.com";
+
+    private static final User toSaveUser = new User.Builder(ID, EMAIL).setName(NAME).setPublicPortfolio().addSelfIntroduction(SELF_INTRODUCTION).addImgUrl(IMG_URL).build();
+    private static User toGetUser;
```
Comment (tlheng) [inline src/test/java/com/google/sps/UserServletTest.java]: why static?

File: src/test/java/com/google/sps/DatastoreUserRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+    private static final String ID_C = "useridC";
+    private static final String EMAIL_C = "user@gmail.com_C";
+    private static final String NAME_C = "username_C";
+    private static final String SELF_INTRODUCTION_C = "I am the user_C";
+    private static final String IMG_URL_C = "/img.com_C";
+
+    private final static User toSaveUser = new User.Builder(ID, EMAIL).setName(NAME).addSelfIntroduction(SELF_INTRODUCTION).addImgUrl(IMG_URL).build(); 
+    private final static User toSaveUser_C = new User.Builder(ID_C, EMAIL_C).setName(NAME_C).addSelfIntroduction(SELF_INTRODUCTION_C).addImgUrl(IMG_URL_C).build(); 
+    private static User toGetUser;
```
Comment (tlheng) [inline src/test/java/com/google/sps/DatastoreUserRepositoryTest.java]: why static?

File: src/main/java/com/google/sps/servlets/UserServlet.java (hunk 6 above)
Comment (bori00) [reply src/main/java/com/google/sps/servlets/UserServlet.java]: Changed.

File: src/test/java/com/google/sps/UserServletTest.java (hunk 7 above)
Comment (bori00) [reply src/test/java/com/google/sps/UserServletTest.java]: Changed.

File: src/test/java/com/google/sps/DatastoreUserRepositoryTest.java (hunk 8 above)
Comment (bori00) [reply src/test/java/com/google/sps/DatastoreUserRepositoryTest.java]: Changed.

**Review by tlheng — APPROVED** [review]

**Pull request merged**